lib/PersonalityProvider.jsm

```diff
@@ -0,0 +1,249 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+const {RemoteSettings} = ChromeUtils.import("resource://services-settings/remote-settings.js", {});
+
+const {actionCreators: ac} = ChromeUtils.import("resource://activity-stream/common/Actions.jsm", {});
+ChromeUtils.defineModuleGetter(this, "perfService", "resource://activity-stream/common/PerfService.jsm");
+
+const {NaiveBayesTextTagger} = ChromeUtils.import("resource://activity-stream/lib/NaiveBayesTextTagger.jsm", {});
+const {NmfTextTagger} = ChromeUtils.import("resource://activity-stream/lib/NmfTextTagger.jsm", {});
+const {RecipeExecutor} = ChromeUtils.import("resource://activity-stream/lib/RecipeExecutor.jsm", {});
+
+ChromeUtils.defineModuleGetter(this, "NewTabUtils",
+  "resource://gre/modules/NewTabUtils.jsm");
+
+/**
+ * V2 provider builds and ranks an interest profile (also called an “interest vector”) off the browse history.
+ * This allows Firefox to classify pages into topics, by examining the text found on the page.
+ * It does this by looking at the history text content, title, and description.
+ */
+this.PersonalityProvider = class PersonalityProvider {
+  constructor(
+    timeSegments,
+    parameterSets,
+    maxHistoryQueryResults,
+    version,
+    scores,
+    v2Params) {
+    this.v2Params = v2Params || {};
+    this.dispatch = this.v2Params.dispatch || (() => {});
+    this.perfStart = perfService.absNow();
+    this.modelKeys = this.v2Params.modelKeys;
+    this.timeSegments = timeSegments;
+    this.parameterSets = parameterSets;
+    this.maxHistoryQueryResults = maxHistoryQueryResults;
+    this.version = version;
+    this.scores = scores || {};
+    this.interestConfig = this.scores.interestConfig;
+    this.interestVector = this.scores.interestVector;
+  }
+
+  async init(callback) {
+    this.interestConfig = this.interestConfig || await this.getRecipe();
+    if (!this.interestConfig) {
+      this.dispatch(ac.PerfEvent({event: "PERSONALIZATION_V2_GET_RECIPE_ERROR"}));
+      return;
+    }
+    this.recipeExecutor = await this.generateRecipeExecutor();
+    if (!this.recipeExecutor) {
+      this.dispatch(ac.PerfEvent({event: "PERSONALIZATION_V2_GENERATE_RECIPE_EXECUTOR_ERROR"}));
+      return;
+    }
+    this.interestVector = this.interestVector || await this.createInterestVector();
+    if (!this.interestVector) {
+      this.dispatch(ac.PerfEvent({event: "PERSONALIZATION_V2_CREATE_INTEREST_VECTOR_ERROR"}));
+      return;
+    }
+
+    this.dispatch(ac.PerfEvent({
+      event: "PERSONALIZATION_V2_TOTAL_DURATION",
+      value: Math.round(perfService.absNow() - this.perfStart),
```

Comment:
Why initialize `this.prefStart` in the constructor? It'll be incorrect if another `init` call gets triggered. I believe a local `perfStart` should be good enough?

Reply:
Yeah this can be updated. I see no issues with this.